cache/cache_reservation_manager.h

```diff
@@ -0,0 +1,81 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// CacheReservationManager is for reserving cache space through
+// inserting/removing dummy entries based on the memory used.
+// This class is not thread-safe.
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "cache/cache_entry_roles.h"
+#include "rocksdb/cache.h"
+#include "rocksdb/slice.h"
+#include "rocksdb/status.h"
+#include "table/block_based/block_based_table_reader.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class CacheReservationManager {
```

Comment:
Do we need to say this class's functions (besides `GetTotalReservedCacheSize()`) are not thread-safe so require external synchronization?

Reply:
Yes - and it seems like the current way of saying "This class is not thread-safe." in https://github.com/facebook/rocksdb/pull/8506/files#diff-b95042a5d09f06814f920b8c334a4f0aff0574c41981d1fc82a93d1c87e774b5R12 does not grasp as much attention as I expected.

Should I try marking non-thread safe/external synchronization needed in front of every function that needs it?

Reply:
> Great progress on generalizing how we charge block cache for outside memory usage! The main thing I wonder about is if we can simplify how we handle reserving one more or one less dummy entry when the target memory falls in the middle.

Thanks! Replied to your thought in https://github.com/facebook/rocksdb/pull/8506/files#r692622024 and will reach out offline tmr to clear up some of my confusion. 

And I will also follow up on the to-do suggestions on the test in [here](https://github.com/facebook/rocksdb/pull/8506/files#r692473085) and function comment in [here](https://github.com/facebook/rocksdb/pull/8506/files#r692600511)

Reply:
Oh sorry... I might've overlooked it because it's grouped with the  copyright notices. That is fine already. Maybe above the class definition is slightly more conventional; I think you don't need to put it above every function.

Reply:
Sure - I can move the description and non thread-safe right above the class definition.

Reply:
> will reach out offline tmr to clear up some of my confusion.

Discussed offline. We now both agree on the approach of updating cache reservation to the ceil(new_mem_used) with respect to the multiples of dummy entry sizes. It eliminates the unnecessary increase/decrease I described in  https://github.com/facebook/rocksdb/pull/8506#discussion_r692511783 and makes it easier to resonate the amount of dummy entries in the cache after update(new_mem_used) since we can now always expect for ceil(new_mem_used).